src/client/texturesource.cpp

```diff
@@ -1848,58 +1856,115 @@ bool TextureSource::generateImagePart(std::string_view part_of_name,
 
 #undef CHECK_DIM
 
-/*
-	Calculate the color of a single pixel drawn on top of another pixel.
 
-	This is a little more complicated than just video::SColor::getInterpolated
-	because getInterpolated does not handle alpha correctly.  For example, a
-	pixel with alpha=64 drawn atop a pixel with alpha=128 should yield a
-	pixel with alpha=160, while getInterpolated would yield alpha=96.
-*/
-static inline video::SColor blitPixel(const video::SColor src_c, const video::SColor dst_c, u32 ratio)
+namespace {
+
+/// Draw src on top of dst
+template <bool overlay, class Colour>
+void blit_pixel(const Colour &src, Colour &dst)
 {
-	if (dst_c.getAlpha() == 0)
-		return src_c;
-	video::SColor out_c = src_c.getInterpolated(dst_c, (float)ratio / 255.0f);
-	out_c.setAlpha(dst_c.getAlpha() + (255 - dst_c.getAlpha()) *
-		src_c.getAlpha() * ratio / (255 * 255));
-	return out_c;
+	if (src.a == 255 || dst.a == 0) {
```

Comment:
have you checked that these extra branches speed it up vs. always performing the calculation?
don't forget that we have utilities for benchmarking: https://github.com/minetest/minetest/tree/master/src/benchmark

Reply:
I have tested it now with example data and the branches made it about twice as fast.
https://github.com/minetest/minetest/pull/14448#issuecomment-1995653086